Remove unnecessary quotes from pipeline + add commit statuses

mokagio opened the pull request:
## Proposed Changes

A formatting-only update in Buildkite pipeline YAML to adopt the (undocumented) @Automattic/apps-infrastructure standard. In particular:

- Omit quotation for strings values unless necessary
- Use Buildkite custom `:name:` syntax for emojis

## Testing Instructions

If the build is green, the change was successful

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors? – N.A.

mokagio committed:
Remove unnecessary quotes from pipeline + use emoji names

mokagio committed:
Use `""` instead of `''` for strings

That's what the Buildkite examples do

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -57,15 +57,15 @@ steps:
+    key: dev-mac
     steps:
-      - label: '🔨 Mac Dev Build - {{matrix}}'
+      - label: ":hammer: Mac Dev Build - {{matrix}}"
```
You might want to add a custom `notify: { github_commit_status: {context: … } }` commit status to go with that change without the GitHub PR checks to be affected…

mokagio replied:
Done.

I also think we agreed to have `notify.github_custom_status` as part of our recommendations, right? I had forgotten about it here.

AliSoftware reviewed (commented):
To be honest, the more I see those changes propagated in various repos, the more I'm starting to change my mind about this convention we agreed on some weeks ago… 😅 

While it helps with consistency to have all emojis in `:name:` format, it's also so much more readable to see the actual emoji in the YAML when possible (i.e. when it's an official one from Unicode that we can insert verbatim in the YAML)…

And all in all I think I value more the visual aspect of the emoji being used verbatim as Unicode character when possible, than the consistency (especially when it happens that not that many of the emojis we use in our labels and logs are non-standard ones…)

---

Also, if we end up doing that change anyway, all the changes in the `label` of any step should probably be accompanied with a `notify: { github_commit_status: {context: … } }` entry to avoid messing up with the GitHub Statuses reported on PRs, right?

mokagio committed:
Revert existing emojis conversion to text

Seems like there's more discussion to be had on this matter,
see https://github.com/Automattic/studio/pull/614#pullrequestreview-2386871890

I don't want what should have been a quick PR to be bogged down on this
granular item. We shall proceed with the discussion elsewhere and
revisit here later.

mokagio committed:
Add `notify.github_commit_status` to all steps in the pipeline

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -57,15 +66,18 @@ steps:
       #- windows
+    notify:
+      - github_commit_status:
+          context: E2E Tests on {{ matrix }}
```
Odd... `{{ matrix }}` is not resolved

<img width="520" alt="image" src="https://github.com/user-attachments/assets/85677f4f-7f5f-4ee7-9093-7ccafe1aeec3">

mokagio replied:
Update, still odd, but at least [expected](https://buildkite.com/docs/pipelines/build-matrix):

![image](https://github.com/user-attachments/assets/36b6cc49-449d-4bb4-9a81-6e1b1663a64a)

mokagio replied:
I experimented with different syntax here https://github.com/wordpress-mobile/WordPress-iOS/pull/23724 (I had forgotten this repo is open source, too) but didn't go anywhere.

I might write in the forums later.

In the meantime, I'll remove the status reporting for the matrix steps.

AliSoftware replied:
Do you know what happens if you set a `notify.github_commit_status` to the step having the `matrix` in practice?
1. Does it only go green once the whole matrix has finished running and each of its jobs went green?
2. Or does each job that is spawn in the matrix emits the same commit status each (e.g. if the first job goes red then status is sent as red and if the next job in the matrix finishes later and ends up reporting green it will re-emit the commit status as green)?

The behavior 2 would not make much sense in practice, but might still be what happens if it's not intended behavior but a big in Buildkite.

The fact that the `{{ matrix }}` in the status name is not resolved and the un-usefulness of behavior 2 makes me think it's likely behavior 1 instead, which would probably make more sense in practice anyway.

So maybe it makes sense to have a single commit status reported for the whole matrix (and you'd just not include `{{ matrix }}` in its `context` because it would reflect the status of the whole matrix)… but even if we decide on that, I'd still want us to verify it's not behavior 2 and a bug before adopting it.

mokagio replied:
> I might write in the forums later.

See https://forum.buildkite.community/t/support-matrix-interpolation-in-github-commit-status-context/4004

mokagio committed:
Wrap `{{matrix}}` usages in `context` in quotes

Will it help with having Buildkite interpolated the matrix value when
writing to GitHub?

See
https://github.com/Automattic/studio/pull/614/files#r1821669523

mokagio committed:
Remove space around `matrix` interpolation

mokagio committed:
Remove GitHub status notification from matrix steps

mokagio commented:
@AliSoftware thanks as always for the input. As you'll have seen from the inline comments,

- I reverted the emoji changes. Let's talk about the standard elsewhere with the rest of the team.
- I adopted the commit status notification, but had to drop it from the steps using a matrix because I couldn't find a way to differentiate them

I think the PR as it is good enough to merge so it doesn't add to my long list of stale PRs. We can keep iterating on this lower priority work time permitting.

wojtekn reviewed (approved):
Thanks for the cleanup.

AliSoftware reviewed (approved):
Looks good :shipit: 

@mokagio I'll let you adjust the GitHub required checks in branch protection settings (`buildkite/studio/lint` -> `Lint`, `buildkite/studio/unit-tests` -> `Unit Tests`) once you merge this, then make the PR in our IaC to set the `publish_commit_status{_per_step}` to `false` 🙂
